File: cpp/data_struct/hash/1-twoSum.py
#!/usr/bin/python3
from typing import List


# Solution1
# 执行用时： 44 ms , 在所有 Python3 提交中击败了 39.48% 的用户
# 内存消耗： 14.9 MB , 在所有 Python3 提交中击败了# 39.90%# 的用户
class Solution1:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        if 0 == len(nums):
            return []
        tmp = dict()
        for ind, n in enumerate(nums):
            kk = tmp.get(target - n)
            if kk is not None:
                return [ind, kk]
            tmp[n] = ind


class Solution2:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        if 0 == len(nums):
            return []
        tmp = dict()
        for ind, n in enumerate(nums):
            if target - n in tmp:
                return [ind, tmp[target - n]]
            tmp[n] = ind


print(Solution2().twoSum([1, 2, 3], 4))

File: cpp/data_struct/hash/1-twosum.cpp
//
// Created by DELL on 2021/7/26.
//

#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int> &nums, int target) {
        unordered_map<int, int> hashtable;
        for (int i = 0; i < nums.size(); ++i) {
            auto it = hashtable.find(target - nums[i]);
            if (it != hashtable.end()) {
                return {it->second, i};
            }
            hashtable[nums[i]] = i;
        }
        return {};
    }
};

int main() {
    Solution tmp;
    vector<int> gg=vector<int>{1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    vector<int> xx = tmp.twoSum(gg, 7);
    for (auto array: xx) {
        std::cout << array << " ";
    }
    std::cout << std::endl;

    return 0;
}
